dbms/src/Flash/Mpp/MinTSOScheduler.cpp

```diff
@@ -178,51 +176,64 @@ void MinTSOScheduler::deleteQuery(
 }
 
 /// NOTE: should not throw exceptions due to being called when destruction.
-void MinTSOScheduler::releaseThreadsThenSchedule(const int needed_threads, MPPTaskManager & task_manager)
+void MinTSOScheduler::releaseThreadsThenSchedule(
+    const String & resource_group_name,
+    int needed_threads,
+    MPPTaskManager & task_manager)
 {
     if (isDisabled())
     {
         return;
     }
 
-    auto updated_estimated_threads = static_cast<Int64>(estimated_thread_usage) - needed_threads;
+    global_estimated_thread_usage -= needed_threads;
+    auto & entry = getOrCreateGroupEntry(resource_group_name);
+    auto updated_estimated_threads = static_cast<Int64>(entry.estimated_thread_usage) - needed_threads;
     RUNTIME_ASSERT(
         updated_estimated_threads >= 0,
         log,
         "estimated_thread_usage should not be smaller than 0, actually is {}.",
         updated_estimated_threads);
 
-    estimated_thread_usage = updated_estimated_threads;
-    GET_METRIC(tiflash_task_scheduler, type_estimated_thread_usage).Set(estimated_thread_usage);
-    GET_METRIC(tiflash_task_scheduler, type_active_tasks_count).Decrement();
+    entry.estimated_thread_usage = updated_estimated_threads;
+    GET_RESOURCE_GROUP_METRIC(tiflash_task_scheduler, type_estimated_thread_usage, resource_group_name)
+        .Set(entry.estimated_thread_usage);
+    GET_RESOURCE_GROUP_METRIC(tiflash_task_scheduler, type_active_tasks_count, resource_group_name).Decrement();
     /// as tasks release some threads, so some tasks would get scheduled.
-    scheduleWaitingQueries(task_manager);
+    scheduleWaitingQueries(entry, task_manager, log);
+    if (entry.active_set.size() + entry.waiting_set.size() == 0 && entry.estimated_thread_usage == 0)
+    {
+        LOG_INFO(log, "min tso scheduler_entry of resouce group {} deleted", resource_group_name);
```

Comment:
Maybe debug is enough.
```suggestion
        LOG_DEBUG(log, "min tso scheduler_entry of resouce group {} deleted", resource_group_name);
```